
syntax = "proto3";

/* DataService: a unified interface for all data operations */
service DataService {
    /* For  ordering, filtering and groupings. */
    rpc ApplyQuery (QueryRequest) returns (QueryResponse);
    /* For individual/bulk inspection of content and stats. */
    rpc GetSamples (SamplesRequest) returns (SamplesResponse);

    /* For overall data statistics. */
    rpc Statistics (StatsRequest) returns (StatsResponse);
    /* For changing anything about a group of Samples: label, tag, presence. */
    rpc EditSample (EditsRequest) returns (EditsResponse);
}

message QueryRequest {
    string query = 1; // The query string to be applied
    bool accumulate = 2; // Whether to accumulate with existing query

    bool is_natural_language = 3; // Whether the query is in natural language
}

message QueryResponse {
    bool success = 1;
    string message = 2;

    int32 number_of_all_samples = 3;
    int32 number_of_samples_in_the_loop = 4; // both eval and train
    int32 number_of_discarded_samples = 5; // both eval and train
}

/*
    Request to get data records along with specified statistics. This is meant
    for fast and efficient data retrieval.
*/
message SamplesRequest {
    int32 start_index = 1;
    int32 records_cnt = 2;

    bool include_transformed_data = 3; // Whether to include model ready data
    bool include_raw_data = 4; // Whether to include raw data

    repeated string stats_to_retrieve = 5; // if empty return all stats

    // string query_to_run_before = 6; // Query to run before fetching data
}

// Represents a single data statistic, such as target, latest_prediction,
// latest_loss, encounters, discarded and any extra fields.
message DataStat {
  string name = 1;
  string type = 2;
  repeated int32 shape = 3;
  repeated float value = 4;
  string value_string = 5;
}

message DataRecord {
  int32 sample_id = 1;
  repeated DataStat data_stats = 2;
}

message SamplesResponse {
  bool success = 1;
  string message = 2;
  repeated DataRecord data_records = 3;
}


message StatsRequest {
  // Runs statistics on the resulting set of samples after running the predicate
  optional string predicate = 1;
}

message StatName2Aggregation {
  string name = 1;
  float value = 2;
}

message StatsResponse {
  repeated StatName2Aggregation statistics = 1;
}


enum SampleEditType {
  OVERRIDE = 0;
  ACCUMULATE = 1;
}

message EditsRequest {
  string stat_name = 1;
  float float_value = 2;
  string string_value = 3;
  bool bool_value = 7;
  SampleEditType type = 4;

  repeated int32 samples_ids = 5;
  repeated string sample_origins = 6;
}

message EditsResponse {
  bool success = 1;
  string message = 2;
}